### ISSUE TITLE
Assertions related to lanes on 5.3 and master

### ISSUE THREAD

**Issue by daniel-j-h** [Project-OSRM/osrm-backend#2730]
Splitting from https://github.com/Project-OSRM/osrm-backend/issues/2706: we're hitting the following lane assertion during extraction on the planet on master but also already on the 5.3 branch:

```
[assert] /mnt/ebs/osrm/2016-07-26-110507/src/extractor/guidance/turn_lane_augmentation.cpp:285
in: osrm::extractor::guidance::lanes::LaneDataVector osrm::extractor::guidance::lanes::handleNoneValueAtSimpleTurn(osrm::extractor::guidance::lanes::LaneDataVector, const Intersection&): connection_count + 1 == lane_data.size()
terminate called without an active exception
```

https://github.com/Project-OSRM/osrm-backend/blob/c7a1576100cb947af0e970644fb3f70dde63a9eb/src/extractor/guidance/turn_lane_augmentation.cpp#L283-L286

I wrote up some code for printing the locations for the breaking intersections in the `fix/2706-debug` branch and deployed it. For the record, here are the first five:

```
>>> 2, 4
 Tags:
	1(none) from: 0 to: 0
	2(straight) from: 1 to: 1
	8(left) from: 2 to: 2
	256(uturn) from: 2 to: 2
>>> Broken Intersection: 
(lon:7.828823, lat:47.976944)
(lon:7.828583, lat:47.975312)
(lon:7.828631, lat:47.97538)
>>> 2, 4
 Tags:
	1(none) from: 0 to: 0
	2(straight) from: 1 to: 2
	8(left) from: 3 to: 3
	256(uturn) from: 3 to: 3
>>> Broken Intersection: 
(lon:7.832026, lat:47.992681)
(lon:7.832197, lat:47.992714)
(lon:7.832428, lat:47.993222)
(lon:7.83195, lat:47.992815)
>>> 2, 4
 Tags:
	1(none) from: 0 to: 0
	2(straight) from: 1 to: 1
	8(left) from: 2 to: 2
	256(uturn) from: 2 to: 2
>>> Broken Intersection: 
(lon:7.809372, lat:48.011799)
(lon:7.809382, lat:48.011674)
(lon:7.809655, lat:48.011597)
(lon:7.809555, lat:48.011808)
>>> 2, 4
 Tags:
	1(none) from: 0 to: 0
	2(straight) from: 1 to: 1
	8(left) from: 2 to: 3
	256(uturn) from: 3 to: 3
>>> Broken Intersection: 
(lon:9.183196, lat:48.776604)
(lon:9.182912, lat:48.776235)
(lon:9.182994, lat:48.776181)
>>> 2, 4
 Tags:
	1(none) from: 0 to: 1
	2(straight) from: 2 to: 2
	8(left) from: 2 to: 3
	256(uturn) from: 4 to: 4
>>> Broken Intersection: 
(lon:-96.884376, lat:32.826898)
(lon:-96.885507, lat:32.82768)
(lon:-96.885885, lat:32.827015)
```

A small extract where we're hitting this would is:
http://download.geofabrik.de/europe/germany/baden-wuerttemberg.html

@danpat could reduce my Cucumber test case even further down to a single lane tag: https://github.com/Project-OSRM/osrm-backend/commit/57ca921dcdf4ef24f58b2c360c1264a98e4480c0

**Comment by danpat** [comment]
Here's the what I've learned about this today:

  1.  The core problem seems to be that this intersection is satisfying `isSimpleIntersection` [here](https://github.com/Project-OSRM/osrm-backend/blob/fix/2706-debug/src/extractor/guidance/turn_lane_handler.cpp#L130), but *it should not be*.
  2.  The reason it is satisfying this is because of [this clause](https://github.com/Project-OSRM/osrm-backend/blob/fix/2706-debug/src/extractor/guidance/turn_lane_handler.cpp#L298-L299) which seems to be a special case designed to handle u-turn lanes.  I don't yet understand why this logic is here, I suspect maybe this is the root cause of the problem.

If I manually prevent (2) from occurring (by commenting it out), then everything seems to work as expected, except a single scenario for 1 test case:  `features/guidance/turn-lanes.feature:84`

This case is very similar to our breaking case, except the lead-in road is 2-way, but the route is forced in a certain direction by using the `bearing` parameter.  The current thing I'm working on understanding is why `intersection[0].entry_allowed` is false for this case - it seems like it should be `true` for a two-way road.

I suspect that the clause for (2) is a workaround for an earlier incorrect setting of `entry_allowed = false`, I'm currently trying to figure out where that's coming from for `turn-lanes.feature:84` and see if the logic there needs to be fixed.

If that's the case and the fix is further upstream, then the entire clause for (2) can be removed and then everything should start working.  I think.

**Comment by danpat** [comment]
This change is a first crack at addressing this:

https://github.com/Project-OSRM/osrm-backend/commit/acb2318a7d0addd0aba52e0110dfe3767a09e52b

I think the additional testing of `at` might be superfluous, but I need to think harder about that to be sure.

I'm still not clear on why an incoming two-way edge has `entry_allowed == false`, I think there is some implied behaviour here that's not obvious.  Hopefully @MoKob can clarify :-)

**Comment by MoKob** [comment]
@danpat u-turns in general are not allowed in OSRM. There has been a long discussion on when to allow/disallow u-turns and summary pretty much is: we don't know when they might be allowed, so we disallow them to be safe. If a specific lane is given to turn-around, the lane-processing enables the turn later on.

**Comment by MoKob** [comment]
#2739 is now accompanied by https://github.com/Project-OSRM/osrm-backend/pull/2753. The extraction of the planet succeeded. This issue should be resolved for good when #2753 is merged.
